File: sections/09_greedy_exchange/problems/c_deadline_schedule/solution.py
"""Reference solution for C. Deadline Schedule."""

import heapq
import sys


def solve_case(tasks: list[tuple[int, int]]) -> int:
    tasks.sort(key=lambda p: (p[1], p[0]))
    total = 0
    chosen: list[int] = []
    for duration, deadline in tasks:
        total += duration
        heapq.heappush(chosen, -duration)
        if total > deadline:
            total += heapq.heappop(chosen)
    return len(chosen)


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        idx += 1
        tasks: list[tuple[int, int]] = []
        for _ in range(n):
            duration = data[idx]
            deadline = data[idx + 1]
            idx += 2
            tasks.append((duration, deadline))
        out.append(str(solve_case(tasks)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/09_greedy_exchange/problems/c_deadline_schedule/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int solve_case(vector<pair<long long, long long>> tasks) {
    sort(tasks.begin(), tasks.end(), [](const auto &a, const auto &b) {
        if (a.second != b.second) return a.second < b.second;
        return a.first < b.first;
    });

    long long total = 0;
    priority_queue<long long> chosen;
    for (auto [duration, deadline] : tasks) {
        total += duration;
        chosen.push(duration);
        if (total > deadline) {
            total -= chosen.top();
            chosen.pop();
        }
    }
    return (int)chosen.size();
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;
        vector<pair<long long, long long>> tasks(n);
        for (auto &[duration, deadline] : tasks) cin >> duration >> deadline;
        cout << solve_case(tasks) << '\n';
    }

    return 0;
}

